File: trunk/jshiny/src/jlo/ioe/ui/Component.java
package jlo.ioe.ui;

import jlo.ioe.Environment;
import jlo.ioe.messaging.MessageService;
import jlo.ioe.ui.behavior.Action;
import jlo.ioe.util.F;

import javax.swing.JComponent;
import javax.swing.KeyStroke;
import javax.swing.SwingUtilities;
import java.awt.Dimension;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;

/**
 * Copyright © 2007 imaginaryday.com (jlo)<br>
 * User: jlowens<br>
 * Date: Apr 7, 2007<br>
 * Time: 2:34:05 PM<br>
 */
class Component<T extends JComponent> implements IComponent<T> {
	protected ActionsHandler actions;
	private CommandInterceptor ci;
	private T comp;

	public Component(T comp) {
		this.comp = comp;
		ci = new CommandInterceptor();
		if (comp instanceof ActionCapable) {
			final ActionCapable ac = (ActionCapable)comp;
			actions = new ActionsHandler() {
				ActionListener al = new ActionListener() {
					public void actionPerformed(ActionEvent actionEvent) {
						MessageService.singleton().publish(new Action(Component.this.comp, actionEvent));
					}
				};
				public void areHandled() {
					ac.addActionListener(al);
				}
				public void areNotHandled() {
					ac.removeActionListener(al);
				}
			};
		} else {
			actions = new ActionsHandler() {
				public void areHandled() {}
				public void areNotHandled() {}
			};
		}
		actions.areNotHandled();
	}

	protected void onSwingThread(final F.lambda<Object> f) {
		SwingUtilities.invokeLater(new Runnable() {
			public void run() {
				f.call();
			}
		});
	}


	public T get() {
		return comp;
	}

	public <T> void update(T val) { }
	public ActionsHandler actions() { return actions; }


	public T comp() {
		return comp;
	}
	protected void comp(T c) { comp = c; }
	public double preferredWidth() { return comp.getPreferredSize().getWidth(); }
	public double preferredHeight() { return comp.getPreferredSize().getHeight(); }
	public Component<T> preferredWidth(double w) { comp.setPreferredSize(new Dimension((int)w,(int)preferredHeight())); return this; }
	public Component<T> preferredHeight(double h) { comp.setPreferredSize(new Dimension((int)preferredWidth(),(int)h)); return this; }
	public Component<T> preferredSize(double w, double h) { comp.setPreferredSize(new Dimension((int)w,(int)h)); return this; }
	public Component<T> setWidth(double w) { comp.setSize(new Dimension((int)w, comp.getHeight())); return this; }
	public Component<T> setHeight(double h) { comp.setSize(new Dimension(comp.getWidth(), (int)h)); return this; }

	private class CommandInterceptor {
		private static final long DIFF = 500;
		private long lastTime = System.currentTimeMillis();

		public CommandInterceptor() {
			System.out.println("added command interceptor: " + comp);
			//new Throwable().printStackTrace();
			comp.addKeyListener(new KeyAdapter() {
				@Override public void keyPressed(KeyEvent e) {
					KeyStroke ks = KeyStroke.getKeyStrokeForEvent(e);
					String s = ks.toString();
					if (s.equals("meta pressed TAB")) {
						Environment.singleton().nextSheet();
						lastTime = 0;
					} else if (s.equals("shift meta pressed TAB")) {
						Environment.singleton().prevSheet();
						lastTime = 0;
					} else if (s.equals("meta pressed META")) {
						System.out.println("meta pressed META");
						long currTime = System.currentTimeMillis();
						if ((currTime - lastTime) < DIFF) {
							Environment.singleton().commandRequested();
							lastTime = 0;
						} else {
							lastTime = currTime;
						}
					}
				}
			});
		}
	}
}

File: trunk/sshiny/src/jlo/ioe/ui/Component.scala
package jlo.ioe.ui;

import java.awt.event._
import javax.swing.JComponent
import javax.swing.border.EmptyBorder
import java.awt.Dimension
import scala.collection.immutable.{HashSet,Set,Map,HashMap}

package behavior {
  case class Action(e:ActionEvent) extends ObservableEvent
}

trait ActionCapable {
  def addActionListener(a:ActionListener)
  def removeActionListener(a:ActionListener)
}

trait ActionsHandler {
  def areHandled
  def areNotHandled
}

trait Component extends JComponent with CommandInterceptor with Observable {
  var actions = if (this.isInstanceOf[ActionCapable]) {
    val o = this.asInstanceOf[ActionCapable]
    new ActionsHandler {
      val _handler = new ActionListener {
	def actionPerformed(e:ActionEvent) = fire(behavior.Action(e))
      }
      def areHandled = o.addActionListener(_handler)
      def areNotHandled = o.removeActionListener(_handler)      
    }
  }
  else
    new ActionsHandler {
      def areHandled = {}
      def areNotHandled = {}
    }
  actions areNotHandled;

  def onSwingThread(f:()=>Unit) {
    javax.swing.SwingUtilities.invokeLater(new Runnable() {
      def run : Unit = {
	f()
      }
    })
  }

  var obsHandlers : Map[Observable,List[EventHandler]] = new HashMap[Observable,List[EventHandler]]()
  def handlers = obsHandlers
  def handlers_=(h:Map[Observable,List[EventHandler]]) = obsHandlers = h

  var obsListeners : Set[Observer] = new HashSet[Observer]()
  def listeners = obsListeners
  def listeners_=(o:Set[Observer]) = obsListeners = o


  // bindings
  def update(v:Any) = {}

  // conveniences
  def margin(t:int,l:int,b:int,r:int) : Component = { setBorder(new EmptyBorder(t,l,b,r)); this }  
  def preferredWidth = getPreferredSize().getWidth()
  def preferredHeight = getPreferredSize().getHeight()
  def preferredWidth(w:double) : Component = { setPreferredSize(new Dimension(w,preferredHeight)); this }
  def preferredHeight(h:double) : Component = { setPreferredSize(new Dimension(preferredWidth,h)); this }  
  def preferredSize(w:double,h:double) : Component = { setPreferredSize(new Dimension(w,h)); this }
  def setWidth(w:double) : Component = { setSize(new Dimension(w,getHeight())); this }
  def setHeight(h:double) : Component = { setSize(new Dimension(getWidth(),h)); this }

  // CONVERSIONS
  implicit def doubleToInt(x:double) : int = x.asInstanceOf[int]

}

class Viewport extends javax.swing.JViewport with Component {}

class Scroller(o:Component) extends javax.swing.JScrollPane(o) with Component {
  //setBorder(null)
  override def createViewport() : javax.swing.JViewport = {
    new Viewport
  }
}

// class Frame(title:String) extends javax.swing.JFrame(title) with Component {}
